builder/openstack/step_allocate_ip.go

```diff
@@ -23,123 +22,149 @@ func (s *StepAllocateIp) Run(_ context.Context, state multistep.StateBag) multis
+		newIP, err := floatingips.Create(networkClient, floatingips.CreateOpts{
+			FloatingNetworkID: floatingNetwork,
+		}).Extract()
+		if err != nil {
+			err := fmt.Errorf("Error creating floating IP from floating network '%s': %s", floatingNetwork, err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
 		}
+
+		instanceIP = *newIP
+		ui.Message(fmt.Sprintf("Created floating IP: '%s' (%s)", instanceIP.ID, instanceIP.FloatingIP))
+		state.Put("floatingip_istemp", true)
 	}
 
-	if instanceIp.IP != "" {
-		ui.Say(fmt.Sprintf("Associating floating IP with server..."))
-		ui.Message(fmt.Sprintf("IP: %s", instanceIp.IP))
-		err := floatingips.AssociateInstance(client, server.ID, floatingips.AssociateOpts{
-			FloatingIP: instanceIp.IP,
-		}).ExtractErr()
+	// Assoctate a floating IP that was obtained in the previous steps.
+	if instanceIP.ID != "" {
+		ui.Say(fmt.Sprintf("Associating floating IP '%s' (%s) with instance port...",
+			instanceIP.ID, instanceIP.FloatingIP))
+
+		portID, err := GetInstancePortID(computeClient, server.ID)
+		if err != nil {
+			err := fmt.Errorf("Error getting interfaces of the instance '%s': %s", server.ID, err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+
+		_, err = floatingips.Update(networkClient, instanceIP.ID, floatingips.UpdateOpts{
+			PortID: &portID,
+		}).Extract()
 		if err != nil {
 			err := fmt.Errorf(
-				"Error associating floating IP %s with instance: %s",
-				instanceIp.IP, err)
+				"Error associating floating IP '%s' (%s) with instance port '%s': %s",
+				instanceIP.ID, instanceIP.FloatingIP, portID, err)
 			state.Put("error", err)
 			ui.Error(err.Error())
 			return multistep.ActionHalt
 		}
 
 		ui.Message(fmt.Sprintf(
-			"Added floating IP %s to instance!", instanceIp.IP))
+			"Added floating IP '%s' (%s) to instance!", instanceIP.ID, instanceIP.FloatingIP))
 	}
 
-	state.Put("access_ip", &instanceIp)
+	state.Put("access_ip", &instanceIP)
 	return multistep.ActionContinue
 }
 
 func (s *StepAllocateIp) Cleanup(state multistep.StateBag) {
 	config := state.Get("config").(Config)
 	ui := state.Get("ui").(packer.Ui)
-	instanceIp := state.Get("access_ip").(*floatingips.FloatingIP)
+	instanceIP := state.Get("access_ip").(*floatingips.FloatingIP)
 
 	// Don't delete pool addresses we didn't allocate
 	if state.Get("floatingip_istemp") == false {
 		return
 	}
 
-	// We need the v2 compute client
-	client, err := config.computeV2Client()
+	// We need the v2 network client
+	client, err := config.networkV2Client()
 	if err != nil {
 		ui.Error(fmt.Sprintf(
-			"Error deleting temporary floating IP %s", instanceIp.IP))
+			"Error deleting temporary floating IP '%s' (%s)", instanceIP.ID, instanceIP.FloatingIP))
 		return
 	}
 
-	if s.FloatingIpPool != "" && instanceIp.ID != "" {
-		if err := floatingips.Delete(client, instanceIp.ID).ExtractErr(); err != nil {
+	if instanceIP.ID != "" {
```

Comment:
This logic changed, was that intentionally and if so why?

Reply:
Nova API Floating IP pool object doesn't exist anymore and it is not needed here.

Users needs to provide an external network to create floating IP with the Neutron API, and we check if they provided it at the [https://github.com/hashicorp/packer/blob/0a0471e0b5983636324ec338a2c3bb4473e5e9e5/builder/openstack/step_allocate_ip.go#L80](step_allocate_ip.go#L80). But with the usage of the Neutron API we don't actually care if it wasn't provided, because we can search for the external network as we do later in: [https://github.com/hashicorp/packer/blob/0a0471e0b5983636324ec338a2c3bb4473e5e9e5/builder/openstack/step_allocate_ip.go#L83](step_allocate_ip.go#L83).

With this info in mind, it is clear that we don't need to know if user provided `FloatingIPNetwork` parameter or not while we are running the `Cleanup` method. We only need to check if the `instanceIP` is not empty so it needs to be cleaned.

Reply:
Oh, but the new logic is not so good.
If we searched for the external network when user didn't provide the `FloatingIPNetwork` parameter, we would always try to associate floating IP to the instance port.

This is wrong because we only need to associate floating IP when user provided the `FloatingIPNetwork` parameter.

I will remove searching of the external network in case of empty `FloatingIPNetwork` and refactor this stuff.